HathorNetwork/hathor-core
# refactor(verification): organization and typing improvements [part 1/9]

### Motivation

This PR is the first one in a series of 9 PRs that finishes the vertex verification refactors. The objective is to completely remove all inheritance from the verification process, removing implicit dependencies between verifiers, streamlining the verification process and improving clarity on the scope/origin of each verification method, all to improve code readability, maintainability, and patching (for the simulator).

This is done incrementally. In previous PRs, we moved the verification out of the model classes. Now, in this series, we'll first change inheritance to composition, and then completely remove the composition too, leaving each verifier with the least possible amount of dependencies.

Also, in this PR, we install the [typing_extensions](https://github.com/python/typing_extensions) dev dependency. This is a project from the official Python organization, and what it does is enable use of new type system features on older Python versions. In other words, it simply allows us to use features from the `typing` module that are not available in all Python versions we support.

Specifically, for this refactor we'll use [`assert_never()`](https://typing.readthedocs.io/en/latest/source/unreachable.html), a function to be used were we would normally use something like `assert False`, but that can be correctly type-checked by mypy. This improves robustness of `match` statements for example, that can be made exhaustive with this. We also use the [`@override`](https://peps.python.org/pep-0698/) annotation in next PRs, improving robustness and safety guarantees of inheritance. We may also find other useful typing features in the future that wouldn't be available otherwise.

### Acceptance Criteria

- Change `BaseTransaction` and all subclasses to use `TxVersion` as the `version` type, instead of `int`.
- Install the `typing_extensions` dev dependency.
- Move `verify_unsigned_skip_pow()` from `TransactionVerifier` to `create_tx.py`. This resolves https://github.com/HathorNetwork/hathor-core/issues/825.
- In `VerificationService`:
  - Change all `isinstance()` checks to `type() is`, which is more restrictive.
  - Change `raise NotImplementedError` usages in `match` statements to `assert_never()`, which can be checked by the static type checker.
  - Remove `validate_vertex_error()`, inlining it `HathorManager`.
- Set `self.verifiers` variable in `test_verification`, updating all usages. This will be useful in the next PRs.
- No behavior should be changed by this PR.

### Checklist

- [x] If you are requesting a merge into `master`, confirm this code is production-ready and can be included in future releases as soon as it gets merged
